### ISSUE TITLE
Lustre DKMS RPM Broken On New ZFS Release

### ISSUE THREAD

**Issue by andrewawagner** [zfsonlinux/zfs#3009]
I am attempting to setup a new Lustre filesystem for testing purposes. My Centos 6 system successfully installs the SPL/ZFS DKMS RMPs. However, install the Lustre DKMS rpm breaks and errors with the following messages:

```
Running Transaction
  Installing : 1:tcl-8.5.7-6.el6.x86_64                                           1/4 
  Installing : expect-5.44.1.15-5.el6_4.x86_64                                    2/4 
  Installing : lustre-dkms-2.4.2-1dkms.el6.noarch                                 3/4 
Non-fatal POSTIN scriptlet failure in rpm package lustre-dkms-2.4.2-1dkms.el6.noarch
Loading new lustre-2.4.2 DKMS files...
Building for 2.6.32-358.23.2.el6.x86_64
Building initial module for 2.6.32-358.23.2.el6.x86_64
configure: error:

Required zfs osd cannot be built due to missing zfs development headers.

Support for zfs can be enabled by downloading the required packages for your
distribution.  See http://zfsonlinux.org/ to determine is zfs is supported by
your distribution.
				
Error! Bad return status for module build on kernel: 2.6.32-358.23.2.el6.x86_64 (x86_64)
Consult /var/lib/dkms/lustre/2.4.2/build/make.log for more information.
```
Broken Configuration ZFS Packages:
zfs-release-1-2.el6.noarch
zfs-0.6.3-1.2.el6.x86_64
libzfs2-0.6.3-1.2.el6.x86_64
zfs-dkms-0.6.3-1.2.el6.noarch

The Lustre DKMS RPM works with ZFS 0.6.3-1 RPMs, but defaults to using the latest ZFS release instead of the older one. Either the Lustre DKSM RPM should have the older ZFS release set as a dependency or be rebuilt to work for the newer ZFS release.

Working Configuration ZFS Packages:
libzfs2-0.6.3-1.el6.x86_64
zfs-0.6.3-1.el6.x86_64
zfs-release-1-2.el6.noarch
zfs-dkms-0.6.3-1.el6.noarch

**Comment by behlendorf** [comment]
Thank for reporting this.  We'll sort it out and refresh the Lustre package.

**Comment by chuckatkins** [comment]
I just ran in to this same problem.  Basically it looks like the dkms.conf for lustre is pointing to source directories for spl and zfs that don't exist. After installing the latest {zfs,spl,lustre}{,-dkms} packages, the following steps should fix it:

```bash
# Patch the broken DKMS config for lustre
sed 's|--with-spl=${dkms_tree}/spl/${SPL_VERSION}/build|--with-spl=${dkms_tree}/spl/${SPL_VERSION}/source|' -i /usr/src/lustre-2.4.2/dkms.conf
sed 's|--with-zfs=${dkms_tree}/zfs/${ZFS_VERSION}/build|--with-zfs=${dkms_tree}/zfs/${ZFS_VERSION}/source|' -i /usr/src/lustre-2.4.2/dkms.conf

# Rebuild and reinstall the kernel module
chroot $CHROOTDIR dkms remove lustre/2.4.2 --all
chroot $CHROOTDIR dkms build   -m lustre -v 2.4.2
chroot $CHROOTDIR dkms install -m lustre -v 2.4.2
```

**Comment by behlendorf** [comment]
@chuckatkins thanks for pointing out the fix.  The repositories should be updated in the next week or two with a lustre-2.5.3 version which includes this fix.  Thanks for being patient.

**Comment by behlendorf** [comment]
The EPEL 6 repositories have been updated with Lustre 2.5.3.

**Comment by sophana** [comment]
Hi

I'm also trying lustre over zfs.
There seem to be a bug remaining in the lustre dkms package or in the documentation.
http://zfsonlinux.org/lustre.html
System is el6.6
lustre install with yum goes ok, but lustre kernel module doesn't look installed
dkms status shows lustre v2.5.3 is added but not installed
When compiling lustre with "dkms install lustre/2.5.3", it shows 
`Error!  Build of osd_ldiskfs.ko failed for: 2.6.32-504.el6.x86_64 (x86_64)`

Looking at the log, it is said

```
Disabling ldiskfs support because complete ext4 source does not exist.

If you are building using kernel-devel packages and require ldiskfs
server support then ensure that the matching kernel-debuginfo-common
and kernel-debuginfo-common-<arch> packages are installed.
```
I tried 2 workarounds:
I removed the last 3 lines in lustre dkms.conf about the ldiskfs module, and it installed.
I tried `yum install kernel-debuginfo-common-x86_64` but dkms won't recompile lustre again, as it seems I can't remove it with dkms remove lustre/2.5.3 --all
but it should be a solution, and maybe the doc should be updated.

**Comment by chuckatkins** [comment]
I can confirm this issue and resolution as well.  I just noticed this recently in a new deployment I had to do.  lustre configures and builds successfully but won't install because the DKMS file is expecting ldiskfs to exist, which it doesn't.  I worked around it the same way, by patching the dkms.conf to remove ldiskfs from the expected modules list and then force a reinstall of the lustre with dkms:
```bash
export LUSTRE_VERSION=2.5.3

# Patch the lustre DKMS build to not include ldiskfs
# Basically removes the last three lines for module entry 29
sed '/\[29\]/d' -i /usr/src/lustre-${LUSTRE_VERSION}/dkms.conf

# Reinstall the lustre modules with the patched dkms config
dkms install -m lustre -v ${LUSTRE_VERSION}
```

**Comment by behlendorf** [comment]
Thanks for the feedback.  If you could open a pull request which fixes the issue against https://github.com/chaos/lustre/commits/2.5.3-zfs I can pull it in when next refreshing the Lustre packages.

**Comment by sophana** [comment]
Do you want to drop ldiskfs support?

**Comment by sophana** [comment]
After some more investigation, it seems that ldiskfs depends on a specific version of e2fsprogs-libs. Shouldn't these rpms  be provided on the same zfs repository?

I tried to compile with these libraries installed from el6.6, and it doesn't seem to work, or maybe its a bug in the lustre compilation system?

If you want to make it work again, simply revert your own commit:
https://github.com/chaos/lustre/commit/b5e36e7e94ccf486b7363c47199c47f6e390aadc

**Comment by behlendorf** [comment]
@sophana yes I did, and that may have been a mistake.  The ldiskfs bits end up being a bit more fragile them the zfs ones.  A kernel update is much more likely to break the ldiskfs build since it depends on the kernel-debuginfo package for ext4 source code.  And it requires a patched e2fsprogs which I've always been uncomfortable with since it will impact your ext2/3/4 too.

So it's not totally clear to me it's a good idea for us to provide the ldiskfs support too.  @sophana @chuckatkins what do you guys think?

**Comment by chuckatkins** [comment]
Even if it builds and configures, will it even work?  I thought the lustre servers using ldiskfs required a patched kernel as well.  One of the side benefits of using zfs for OSTs and MDTs was that you could use the stock kernel from your distro.

It seems unlikely to me that you'd want both ldiskfs and zfs volumes on the same machine so I guess as a convenience factor I'd opt for the zfs-only build for now.

**Comment by behlendorf** [comment]
_If_ it builds it should work, but it will be more fragile.

**Comment by sophana** [comment]
I tried lustre some time ago on ldiskfs. Using block devices was not very convenient.
Yes the kernel was specific at that time (for servers)

Now, lustre over zfs is so much flexible! Will never go back!
I haven't tried some scale performance tests yet, but it seems to work great.

Ability to create lustre filesystems over existing zfs pools is not documented anywhere, but it seems to work just great, and it should be documented. I don't really understand why the lustre over zfs example is over block devices. Is there a problem using existing pools?

Conclusion: dropping ldiskfs should be ok.
